hydra-node/src/Hydra/Network/Reliability.hs

```diff
@@ -0,0 +1,294 @@
+
+instance Arbitrary ReliabilityLog where
+  arbitrary = genericArbitrary
+
+-- | Middleware function to handle message counters tracking and resending logic.
+--
+-- '''NOTE''': There is some "abstraction leak" here, because the `withReliability`
+-- layer is tied to a specific structure of other layers, eg. be between
+-- `withHeartbeat` and `withAuthenticate` layers.
+--
+-- NOTE: better use of Vectors? We should perhaps use a `MVector` to be able to
+-- mutate in-place and not need `zipWith`
+withReliability ::
+  (MonadThrow (STM m), MonadThrow m, MonadAsync m) =>
+  -- | Tracer for logging messages.
+  Tracer m ReliabilityLog ->
+  -- | Our own party identifier.
+  Party ->
+  -- | Other parties' identifiers.
+  [Party] ->
+  -- | Underlying network component providing consuming and sending channels.
+  NetworkComponent m (Authenticated (ReliableMsg (Heartbeat msg))) (ReliableMsg (Heartbeat msg)) a ->
+  NetworkComponent m (Authenticated (Heartbeat msg)) (Heartbeat msg) a
+withReliability tracer me otherParties withRawNetwork callback action = do
+  ackCounter <- newTVarIO $ replicate (length allParties) 0
+  sentMessages <- newTVarIO IMap.empty
+  seenMessages <- newTVarIO $ Map.fromList $ (,0) <$> toList allParties
+  resendQ <- newTQueueIO
+  ourIndex <- findPartyIndex me
+  let resend = writeTQueue resendQ
+  withRawNetwork (reliableCallback ackCounter sentMessages seenMessages resend) $ \network@Network{broadcast} -> do
+    withAsync (forever $ atomically (readTQueue resendQ) >>= broadcast) $ \_ ->
+      reliableBroadcast ourIndex ackCounter sentMessages network
+ where
+  allParties = fromList $ sort $ me : otherParties
+
+  reliableBroadcast ourIndex ackCounter sentMessages Network{broadcast} =
+    action $
+      Network
+        { broadcast = \msg ->
+            case msg of
+              Data{} -> do
+                ackCounter' <- atomically $ do
+                  acks <- readTVar ackCounter
+                  let newAcks = constructAcks acks ourIndex
+                  writeTVar ackCounter newAcks
+                  modifyTVar' sentMessages (insertNewMsg msg)
+                  readTVar ackCounter
+
+                traceWith tracer (BroadcastCounter ourIndex ackCounter')
+                broadcast $ ReliableMsg ackCounter' msg
+              Ping{} -> do
+                acks <- readTVarIO ackCounter
+                traceWith tracer (BroadcastPing ourIndex acks)
+                broadcast $ ReliableMsg acks msg
+        }
+
+  reliableCallback ackCounter sentMessages seenMessages resend (Authenticated (ReliableMsg acks msg) party) = do
+    if length acks /= length allParties
+      then ignoreMalformedMessages
+      else do
+        partyIndex <- findPartyIndex party
+        (shouldCallback, messageAckForParty, knownAckForParty, knownAcks) <- atomically $ do
+          let messageAckForParty = acks ! partyIndex
+          knownAcks <- readTVar ackCounter
+          let knownAckForParty = knownAcks ! partyIndex
```

Comment:
`(!?)` is the safe vector accessor that returns a `Maybe a`, we should use that and check well formedness of the data upon entry in the `callback` function. We could use the actual number of parties as a type-level index because it is statically known at compile time to be equal to the length of `otherParties` plus 1, but that would lead us into non-boring haskell territory :)